Add initial draft of Auth GEP 1494 (#3500)
kubernetes-sigs/gateway-api: geps/gep-1494/index.md

# GEP-1494: HTTP Auth in Gateway API

* Issue: [#1494](https://github.com/kubernetes-sigs/gateway-api/issues/1494)
* Status: Provisional

(See status definitions [here](/geps/overview/#gep-states).)

## TLDR

> Provide a method for configuring **Gateway API implementations** to add HTTP Authentication for north-south traffic. The method may also include Authorization config if practical.

Reviewer: Hmm.. Are these auth* applicable only to inbound traffic, or do they cover both ingress and egress?

Author: Theoretically, it could cover both, depending on how you orient the Gateway, but I've added something to say this is only ingress for now. Good point, thanks!

## Goals

* A way to configure a Gateway Implementation to perform Authentication (at least), with optional Authorization on behalf of Ana the Application Developer.

* A way for Chihiro the Cluster Admin to configure a default Authentication and/or Authorization config for some set of HTTPRoutes.

> * Optionally, a way for Ana to have the ability to disable Authentication and/or Authorization for specific routes when needed, allowing certain routes to not be protected.

Reviewer 2: Does this imply that Ana has the ability to override the cluster-admin configuration? I assume some use cases for this functionality might involve testing, but I’m curious to hear about other potential use cases. My concern is that an app developer, who may not have a strong understanding of authentication, would have the ability to override cluster-admin (or security admin) defaults.

Reviewer: > Does this imply that Ana has the ability to override the cluster-admin configuration? I assume some use cases for this functionality might involve testing, but I’m curious to hear about other potential use cases. My concern is that an app developer, who may not have a strong understanding of authentication, would have the ability to override cluster-admin (or security admin) defaults.

I think some public pages, like a status page or help page, might not need authn? 

It could be more flexible if the Gateway API allowed an HTTPRoute to either override or inherit settings from the Gateway under a hierarchical control model. This flexibility would be especially useful in cases where roles overlap, such as in smaller organizations where Ana and Chihiro might be the same person.

Author: Yes, the most likely use case here is for healthchecks, or being able to say "the whole website needs auth, except for `/public`" or something.

Reviewer 3: From my experience, cluster admins often lack a clear understanding of the deployed applications and their paths unless developers explicitly provide this information. In this context, the role of the cluster admin would be to enable default authentication globally for all routes. Developers, like Ana, can then override this default authentication at the httproute level or for specific routes within the httproute. WDYT?

Author: Yes, this is roughly what I'm thinking here. A default, not an override, although I'm not sure we'll be using Policy Attachment to do this or not yet.

Author: I updated this wording.

|[Istio](https://istio.io/latest/docs/tasks/security/authorization/authz-custom/)|[Envoy + istio (peer authN)](https://istio.io/latest/docs/concepts/security/#mutual-tls-authentication)|[JWT (Request authN)](https://istio.io/latest/docs/concepts/security/#authentication)|
|[Envoy Gateway](https://gateway.envoyproxy.io/docs/tasks/security/ext-auth/)| [Envoy](https://gateway.envoyproxy.io/docs/tasks/security/ext-auth/#http-external-authorization-service) | [Envoy(JWT)](https://gateway.envoyproxy.io/docs/tasks/security/jwt-authentication/), [Basic](https://gateway.envoyproxy.io/docs/tasks/security/basic-auth/) |
|[Consul](https://developer.hashicorp.com/consul/docs/connect/gateways/api-gateway/secure-traffic/verify-jwts-k8s)|[Envoy](https://developer.hashicorp.com/consul/docs/connect/proxies/envoy-extensions/configuration/ext-authz)| [JWT](https://developer.hashicorp.com/consul/docs/connect/gateways/api-gateway/secure-traffic/verify-jwts-k8s#use-jwts-to-verify-requests-to-api-gateways-on-kubernetes)|

## API

(... details, can point to PR with changes)

## Conformance Details

> (This section describes the names to be used for the feature or
> features in conformance tests and profiles.
> 
> These should be `CamelCase` names that specify the feature as
> precisely as possible, and are particularly important for
> Extended features, since they may be surfaced to users.)

Reviewer 2: ```suggestion
(from https://github.com/kubernetes-sigs/gateway-api/blob/main/geps/gep-2162/index.md#standardize-features-and-conformance-tests-names)

#### Feature Names

Every feature should:

1. Start with the resource name. i.e HTTPRouteXXX
2. Follow the PascalCase convention. Note that the resource name in the string should come as is and not be converted to PascalCase, i.e HTTPRoutePortRedirect and not HttpRoutePortRedirect.
3. Not exceed 128 characters.
4. Contain only letters and numbers

### Conformance tests 

Conformance tests file names should try to follow the the `pascal-case-name.go` format.
For example for `HTTPRoutePortRedirect` - the test file would be `httproute-port-redirect.go`.

Treat this guidance as "best effort" because we might have test files that check the combination of several features and can't follow the same format.

In any case, the conformance tests file names should be meaningful and easy to understand.
```

Reviewer 2: created https://github.com/kubernetes-sigs/gateway-api/pull/3502 as well to fix the template

# GEP-1494: HTTP Auth in Gateway API

* Issue: [#1494](https://github.com/kubernetes-sigs/gateway-api/issues/1494)
* Status: Provisional

(See status definitions [here](/geps/overview/#gep-states).)

## TLDR

Provide a method for configuring **Gateway API implementations** to add HTTP Authentication for north-south traffic. The method may also include Authorization config if practical. At the time of writing, this authentication is only for ingress traffic to the cluster.

## Goals

(Using the [Gateway API Personas](https://gateway-api.sigs.k8s.io/concepts/roles-and-personas/))

* A way to configure a Gateway Implementation to perform Authentication (at least), with optional Authorization on behalf of Ana the Application Developer.

> * A way for Chihiro the Cluster Admin to configure a default Authentication and/or Authorization config for some set of HTTPRoutes.

Reviewer 4: Since we're saying this is exclusively for north-south traffic, I think I'd rather attach this config at the Gateway level. Maybe it would be safer to say something like this:
```suggestion
* A way for Chihiro the Cluster Admin to configure a default Authentication and/or Authorization config for some set of HTTP or GRPC matching criteria.
```

> * Optionally, a way for Ana to have the ability to disable Authentication and/or Authorization for specific routes when needed, allowing certain routes to not be protected.

Reviewer 5: As I understand it, the current proposal *only* allows for "add auth to this website, except for this route". Could this be expanded to have some kind of rule-based authentication selection/bypass mechanism? I'd love to have the ability to choose a specific authentication mechanism (or bypass auth entirely) on a per-route basis, based on multiple factors (e.g. client IP, user agent).

This would make the Gateway API an amazing Layer 7 firewall, but I'm not sure if the project wants to support these kinds of capabilities; I saw [this proposal](https://github.com/kubernetes-sigs/gateway-api/blob/99a3934c6bc1ce0874f3a4c5f20cafd8977ffcb4/geps/gep-735/index.md?plain=1#L12) was closed partly because the feature "operated too much like a WAF/firewall".

Proposed user stories:

```md
* As Ana the Application Developer, I'm maintaining a legacy application that doesn't have any existing authentication mechanisms. I wish to enforce SSO when a user accesses the application through a browser. However, I wish to bypass authentication when the user accesses the application through the mobile app (based on the user agent), as it can't handle the SSO flow.
* As Ana the Application Developer, I wish to be able to use different authentication mechanisms for the same route based on the context of the request. I wish to use SSO when a user accesses my API through a browser, and use JWT when using my mobile app,
* As Chihiro the Cluster Admin, I wish to block undesirable User Agents or IP ranges (spiders, scrapers etc.) from accessing any services exposed by Gateway API inside my cluster.
* As Chihiro the Cluster Admin, I wish to be able to bypass AuthN when requests come in from a trusted, controlled IP range, so external services can access APIs running on my cluster without issue
```

Author: I agree that all of those are interesting use cases, but I think that we need to handle the simpler (enable/disable) case first.

**Authentication (AuthN for short)** refers to proving the requesting party's identity in some fashion.

**Authorization (AuthZ for short)** refers to making decisions about what a party can access _based on their Authenticated identity_.

In this document, we'll use Auth* to represent "AuthN and maybe AuthZ as well".

This capability is useful for both Ana (who wants to have something for AuthN and AuthZ that ensures security without it needing to be built into her app), and Chihiro (who wants to be able to ensure that the platform as a whole conforms to some given level of security).

### Common authentication methods

Before discussing any proposed Auth* solution, it's important to discuss some AuthN methods that are often used in securing modern applications.

#### Basic HTTP Auth

In Basic HTTP Auth, a server asks for authentication as part of returning a `401` status response, and the client includes an `Authorization` header that includes a Base64-encoded username and password.

> Because the password is only _encoded_ and not _encrypted_, Basic Auth is totally unsafe when used outside of an encrypted session (like a HTTPS connection).

Reviewer 6: While the "raw" passwords in basic auth have additional issues (long lived, impersonating human on login pages), JWT and afaik OAUTH/OIDC are not generally not secure against replay attacks.

I.e. besides mTLS (which enforces confidentiality) all mechanisms are insecure in plaintext messages.

I think it's best to move the notion of requiring encryption for safe usage out of Basic Auth and potentially add a note here that basic auth has an additional issue with long-lived (potentially higher power) tokens being exchanged.

Author: I agree with your points here, but the intent here was to say "nobody in their right mind uses Basic Auth outside of _at least_ an encrypted connection" (in a specifcation kind of way of course).

I'm not trying to document _all_ of the possible issues with every one of these, just to give a very basic overview so that everyone reading this who might not have all the context can get _some_ background.

I updated the wording here, PTAL.

Reviewer 6: I do get why it's here. The suggestion is mostly there to avoid someone taking the omission on other methods as sign that they are ok in an unencrypted context, when they should still be avoided.

OTOH, this isn't a tutorial about auth. And likely not front and center and the chances to trip up someone are low enough.

Author: It's a fair point, but I'd like to keep it in mind and get the current state merged so we can start working on the API design.

Basic auth is defined in [RFC-7617](https://datatracker.ietf.org/doc/html/rfc7617).

> #### TLS Client Certificate Authentication

Reviewer 4: May be worth referring to https://gateway-api.sigs.k8s.io/geps/gep-91/ as this is already in progress.

TLS includes the possibility of having both the client and server present certificates for the other party to validate. (This is often called "mutual TLS", but is distinct from the use of that term in Service Mesh contexts, where it means something more like "mutual TLS with short-lifetime, automatically created and managed dynamic keypairs for both client and server").

> In this case, the server authenticates the client based on the client presenting a certificate that's signed by an authority that's also trusted by the server's trust chain. Some implementations also allow details about the certificate to be passed through to backend clients, to be used in authorization decisions.

Reviewer 6: I think the "also" is misplaced, since it's about the server also verifying a client, not about the client's cert being trusted by the same CA as ??.
The "trust chain" seems weird as well. Generally we have trust roots on a server, and the chain is presented by the client.
Since the client side authentication process of the server is implied prior knowledge, I think this can be condensed a bit.

I suggest:

`In this case, the server also authenticates the client, based on the certificate chain presented by the client. Some implementations also allow details about the certificate to be passed through to backend clients, to be used in authorization decisions`

Author: Okay, thanks.

- a common authority issues keypairs to clients and servers
- clients include JWTs with requests that identify themselves and their intended uses
- servers unwrap encryption and validate signatures of these JWTs to validate the chain of trust
- servers use the authentication details to make authorization decisions.

JWT is defined in [RFC-75199](https://datatracker.ietf.org/doc/html/rfc7519).

#### Oauth2 and OIDC

Oauth2 is an _authorization framework_, which allows clients and servers to define ways to perform authentication and authorization in as secure a way as possible. It extensively uses TLS for encryption, and involves a third-party handling the authorization handshake with a client, which the third-party then provides to the server.

Open ID Conect (OIDC) is a protocol based on the OAuth 2 framework, that allows Users to talk to Identity Providers (IDPs), on behalf of a Relying Party (RP), and have the IDP give the user an Identity Token (which the User's browser can then provide as Authentication to the Relying Party), and also allows the RP to request Claims about the User, which can be used for Authorization.

## Auth* User Stories

> * As Ana the Application Developer, I wish to be able to configure that some or all of my service exposed via Gateway API requires Authentication, and ideally to be able to make Authorization decisions about _which_ authenticated clients are allowed to access.

Reviewer 6: This might leak into the "API" phase, but I think there's 2 levels to this, which are both worth an explicit mention

* The API proposed implementation provides enough flexibility to integrate with an authorization mechanism and protect resources entirely in the gateway
* The API allows to inject information about the authentication result into the requests and allows bakend application to make authorization decisions based on this.

Author: I agree with this point, and also that we're probably better to address this in the API phase. I'm going to leave this for now, and we'll come back to it in the next PR.

Reviewer 7: (could compile comments like this that need to be accounted for later, but not as a part of this PR, as a TODO list right within the document)

Author: (The API that we have in mind so far - basically promoting Envoy's ext_auth API to be a Gateway API one) does have the facility to handle both of these use cases. But we can't really start talking about until this PR merges.)

Reviewer 8: Regarding:

> ... ideally to be able to make Authorization decisions ...
   
> ... allows backend application to make authorization decisions based on this.
   
Wouldn't authorization (i.e. granting users' privileges, or otherwise producing credentials) normally be handled by an identity provider? At least in all except the basic auth case.

Reviewer 9: > Wouldn't authorization (i.e. granting users' privileges, or otherwise producing credentials) normally be handled by an identity provider?

I think it depends - in OAuth/JWT cases yea that may be common, for other scenarios like mesh AuthZ policies, the infrastructure/identity provider just handles distribution of the authentication credentials (certificates used for mTLS in many implementations), and the actual policies are typically written as CRDs in the cluster by an platform engineer or developer, and enforced in the dataplane near the destination app, typically by a proxy.

(But this is getting _way_ ahead of the scope we want to immediately tackle to start.)

* As Chihiro the Cluster Admin, I wish to be able to configure default Authentication settings (at least), with an option to enforce Authentication settings (preferable but not required) for some set of services exposed via Gateway API inside my cluster.

> * More User Stories welcomed here!

Reviewer 6: As Ana the Application Developer, I wish to be able to redirect users to a login page when they lack authentication, while unauthenticated API access gets the proper 40x response.

^ to make it clear that we need to handle "human" clients (browsers) slightly different to API consumers due to 30X/40X conventions.

# GEP-1494: HTTP Auth in Gateway API

* Issue: [#1494](https://github.com/kubernetes-sigs/gateway-api/issues/1494)
* Status: Provisional

(See status definitions [here](/geps/overview/#gep-states).)

## TLDR

Provide a method for configuring **Gateway API implementations** to add HTTP Authentication for north-south traffic. The method may also include Authorization config if practical. At the time of writing, this authentication is only for ingress traffic to the cluster.

## Goals

(Using the [Gateway API Personas](https://gateway-api.sigs.k8s.io/concepts/roles-and-personas/))

* A way to configure a Gateway Implementation to perform Authentication (at least), with optional Authorization on behalf of Ana the Application Developer.

> * A way for Chihiro the Cluster Admin to configure a default Authentication and/or Authorization config for some set of HTTP or GRPC matching criteria.

Author: Given how much discussion we've generated with the current scope, I've been very reluctant to consider adding any more.

I'll put into "deferred goals" thinking about how this may be extended for other protocols.

> * Optionally, a way for Ana to have the ability to disable Authentication and/or Authorization for specific routes when needed, allowing certain routes to not be protected. This would probably need to work something like a default enabling at Gateway level, that can be specifically set at lower levels, but further design is TBD.

Reviewer 7: I think it might be helpful to draw out some motivation which explains this particular goal. Then, if we're going to provide a mechanism for the developer to override something [that is security related] set by the cluster administrator, we may want to consider flags to enable this override that Chihiro has to explicitly turn on.

Reviewer 9: This feels less like flags and more like the defaults/overrides of the policy attachment model.

Some example HTTP endpoints that may be opted out of a default AuthN policy could be `/login` (for flows other than client-cert mTLS, unauthenticated human users will often need to get redirected _somewhere_ to authenticate and acquire a credential), or `/health` (some service health-checking mechanisms may prefer to avoid the overhead of authentication for performance reasons).

Reviewer 7: Interesting :thinking:  An explanation of the motivation seems helpful in any case and maybe some elaboration here on the goal since apparently we two had somewhat different interpretations of it.

Author: I've updated with some other wording, PTAL. In the interest of time, we could also move this to Deferred Goals, and get the basic setup handled first (with an eye to being able to handle something like this in the future).

This capability is useful for both Ana (who wants to have something for AuthN and AuthZ that ensures security without it needing to be built into her app), and Chihiro (who wants to be able to ensure that the platform as a whole conforms to some given level of security).

### Common authentication methods

Before discussing any proposed Auth* solution, it's important to discuss some AuthN methods that are often used in securing modern applications.

#### Basic HTTP Auth

In Basic HTTP Auth, a server asks for authentication as part of returning a `401` status response, and the client includes an `Authorization` header that includes a Base64-encoded username and password.

Use of passwords in this way is very vulnerable to replay attacks (since if you get the password, you can use
it as many times as you like), which is why Basic Auth is generally used for lower-security use cases.

However, even in the lowest-security use cases, using TLS to at least prevent man-in-the-middle password interception is very necessary.

Basic auth is defined in [RFC-7617](https://datatracker.ietf.org/doc/html/rfc7617).

#### TLS Client Certificate Authentication

> TLS includes the possibility of having both the client and server present certificates for the other party to validate. (This is often called "mutual TLS", but is distinct from the use of that term in Service Mesh contexts, where it means something more like "mutual TLS with short-lifetime, automatically created and managed dynamic keypairs for both client and server").

Reviewer 7: :+1:

## Auth* User Stories

* As Ana the Application Developer, I wish to be able to configure that some or all of my service exposed via Gateway API requires Authentication, and ideally to be able to make Authorization decisions about _which_ authenticated clients are allowed to access.
* As Ana the Application Developer, I wish to be able to redirect users to a login page when they lack authentication, while unauthenticated API access gets the proper 40x response.
* As Chihiro the Cluster Admin, I wish to be able to configure default Authentication settings (at least), with an option to enforce Authentication settings (preferable but not required) for some set of services exposed via Gateway API inside my cluster.
* More User Stories welcomed here!

## Currently implemented Auth mechanisms in implementations

> Many Gateway API implementations have implemented some method of configuring Auth* that are currently Implementation Specific.

Reviewer 9: At least for the Istio examples, these aren't actually using Gateway API implementation-specific extensibility mechanisms (HTTPRoute filters, etc), but rather custom annotations and CRDs. It may be worth clarifying this statement a bit to better reflect if these are intended to be Gateway API well-known or ad-hoc extension examples, or just capabilities of implementations which also support Gateway API configuration.

|[ingress-nginx](https://kubernetes.github.io/ingress-nginx/examples/customization/external-auth-headers/)|[httpbin](https://httpbin.org) ([Basic](https://kubernetes.github.io/ingress-nginx/examples/auth/external-auth/), [OAuth](https://kubernetes.github.io/ingress-nginx/examples/auth/oauth-external-auth/))|[Basic](https://kubernetes.github.io/ingress-nginx/examples/auth/basic/), [Client Certificate](https://kubernetes.github.io/ingress-nginx/examples/auth/client-certs/)|
|[Envoy](https://www.envoyproxy.io/docs/envoy/latest/api-v3/extensions/filters/http/ext_authz/v3/ext_authz.proto)|[External Authorization server (ext_authz filter)](https://www.envoyproxy.io/docs/envoy/latest/api-v3/extensions/filters/http/ext_authz/v3/ext_authz.proto) |[JWT](https://www.envoyproxy.io/docs/envoy/latest/api-v3/extensions/filters/http/jwt_authn/v3/config.proto)|
|[Contour](https://projectcontour.io/docs/1.24/config/client-authorization/)|[Envoy](https://projectcontour.io/docs/1.24/config/client-authorization/)|[Envoy(JWT)](https://projectcontour.io/docs/1.24/config/jwt-verification/)|

> |[Istio](https://istio.io/latest/docs/tasks/security/authorization/authz-custom/)|[Envoy + istio (peer authN)](https://istio.io/latest/docs/concepts/security/#mutual-tls-authentication)|[JWT (Request authN)](https://istio.io/latest/docs/concepts/security/#authentication)|

Reviewer 9: ```suggestion
|[Istio](https://istio.io/latest/docs/tasks/security/authorization/)|[mutual TLS ingress gateway](https://istio.io/latest/docs/tasks/traffic-management/ingress/secure-ingress/#configure-a-mutual-tls-ingress-gateway), [External Authorization](https://istio.io/latest/docs/tasks/security/authorization/authz-custom/)|[JWT (RequestAuthentication)](https://istio.io/latest/docs/reference/config/security/request_authentication/)|
```
I think the mutual TLS ingress gateway docs are a more appropriate link for external auth than PeerAuthentication, which is focused on service-to-service configuration and out of scope for this GEP. Moved the External Authorization link to be more explicit, that feature is more specific than general external-facing (ingress) authentication. Updated link for JWT RequestAuthentication to the full doc page.

I'm unsure if it may also be worth referencing the [JWT claim-based routing](https://istio.io/latest/docs/tasks/security/authentication/jwt-route/) capabilities, which AFAIK are only configurable via an Istio VirtualService, or if any more specific documentation is worth linking to regarding External Authorization for an ingress gateway or for services exposed via an ingress gateway.

@howardjohn @keithmattix if I missed any documented functionality that may be relevant

Author: Updated, thanks.

# GEP-1494: HTTP Auth in Gateway API

* Issue: [#1494](https://github.com/kubernetes-sigs/gateway-api/issues/1494)
* Status: Provisional

(See status definitions [here](/geps/overview/#gep-states).)

## TLDR

Provide a method for configuring **Gateway API implementations** to add HTTP Authentication for north-south traffic. The method may also include Authorization config if practical. At the time of writing, this authentication is only for ingress traffic to the cluster.

## Goals

(Using the [Gateway API Personas](https://gateway-api.sigs.k8s.io/concepts/roles-and-personas/))

* A way for Ana the Application Developer to configure a Gateway API implementation to perform Authentication (at least), with optional Authorization.

* A way for Chihiro the Cluster Admin to configure a default Authentication and/or Authorization config for some set of HTTP or GRPC matching criteria.

> * Optionally, a way for Ana to have the ability to disable Authentication and/or Authorization for specific routes when needed, allowing certain routes to not be protected. This would probably need to work something like a default enabling at Gateway level, that can be specifically set at lower levels, but further design is TBD. This goal comes from the relatively-common desire for Chihiro to be able to set reasonably-secure defaults, and for Ana or others to be able to _disable_ for specific paths for purposes of healthchecking. The fact that this is relatively undefined is why this goal is _optional_.

Reviewer 4: Or reading further, maybe this is a "deferred goal"

> * Optionally, a way for Ana to have the ability to disable Authentication and/or Authorization for specific routes when needed, allowing certain routes to not be protected. This would probably need to work something like a default enabling at Gateway level, that can be specifically set at lower levels, but further design is TBD. This goal comes from the relatively-common desire for Chihiro to be able to set reasonably-secure defaults, and for Ana or others to be able to _disable_ for specific paths for purposes of healthchecking. The fact that this is relatively undefined is why this goal is _optional_.

Author: I think Stretch Goal is better, because it would be easy to accidentally prohibit this unless we think about it a bit at API design time.

## Non-Goals

* Handling all possible authentication and authorization schemes. Handling a (preferably large) subset of authentication and authorization is acceptable.

## Deferred Goals

* (Not performed during the Provisional Phase) Defining the API or configuration to be used.

> * Handling GRPCRoute (We will handle plain HTTP first)

Reviewer 4: Slight change to avoid any suggestion on whether this should be attached at a Route or Gateway level.
```suggestion
* Handling GRPC (We will handle plain HTTP first)
```

Open ID Conect (OIDC) is a protocol based on the OAuth 2 framework, that allows Users to talk to Identity Providers (IDPs), on behalf of a Relying Party (RP), and have the IDP give the user an Identity Token (which the User's browser can then provide as Authentication to the Relying Party), and also allows the RP to request Claims about the User, which can be used for Authorization.

Usually, the Identity Token is delivered using JWT, although that is not required.

## Auth* User Stories

* As Ana the Application Developer, I wish to be able to configure that some or all of my service exposed via Gateway API requires Authentication, and ideally to be able to make Authorization decisions about _which_ authenticated clients are allowed to access.
* As Ana the Application Developer, I wish to be able to redirect users to a login page when they lack authentication, while unauthenticated API access gets the proper 40x response.
* As Chihiro the Cluster Admin, I wish to be able to configure default Authentication settings (at least), with an option to enforce Authentication settings (preferable but not required) for some set of services exposed via Gateway API inside my cluster.

> * More User Stories welcomed here!

Reviewer 10: To add a a few (feel free chose or ignore):

1. As Ana the Application Developer, I want to integrate my Gateway API with an external authorization service like Okta or Auth0, so that I can leverage existing enterprise identity providers without implementing custom authentication logic.

3. As Ana the Application Developer, I want to validate JWTs against a configurable JWKS endpoint so that I can use tokens issued by different identity providers across different environments (development, staging, production).

4. As Ana the Application Developer, I need to simultaneously support multiple authentication methods with clear precedence rules (like mTLS for service clients and JWT for user requests), ensuring each client type can authenticate appropriately while maintaining consistent identity determination when multiple credentials are present.

I got more ideas around monitoring failures, caching JWT validation results, fallback, but i'll save those for later.
